### ISSUE TITLE
Addittional feature parity with vim-surround

### ISSUE THREAD

**Issue by ibhagwan** [echasnovski/mini.nvim#84]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.surround

### Description

Hi @echasnovski,

Thank you so much for writing this, your work is amazing, your code is beautiful and very well thought out and documented, I've yet to explore anything other than `mini.surround` but the amount of functionality in your framework is mind blowing and the mini name does it injustice, surround by itself could be it's own plugin/ecosystem with recipes and configurations, etc.

I certainly didn't expect something with `mini` in it's name to have so much functionality, supporting anything I could imagine with custom lua handlers/regexes, coming from vim-surround I had no idea how useful the highlighting feature  and  `cover_or_next` could be.

The plugin as is is practically perfect for my needs so feel free to close this issue, this is just something that came up in my testing while migrating from `vim-surround` and is definitly in the "nice to have" category.

<details>
<summary>I use surround-like configuration/keybinds:</summary>

```lua
surround.setup({
  -- vim-surround style mappings
  -- left brackets add space around the text object
  -- 'ysiw('    foo -> ( foo )
  -- 'ysiw)'    foo ->  (foo)
  custom_surroundings = {
    ['('] = { output = { left = '( ', right = ' )' } },
    ['['] = { output = { left = '[ ', right = ' ]' } },
    ['{'] = { output = { left = '{ ', right = ' }' } },
    ['<'] = { output = { left = '< ', right = ' >' } },
    S = {
      -- lua bracketed string mapping
      -- 'ysiwS'  foo -> [[foo]]
      input = { find = '%[%[.-%]%]', extract = '^(..).*(..)$' },
      output = { left = '[[', right = ']]' },
    },
  },
  mappings = {
    add = 'ys',
    delete = 'ds',
    find = '',
    find_left = '',
    highlight = 'gs',     -- hijack 'gs' (sleep) for highlight
    replace = 'cs',
    update_n_lines = '',  -- bind for updating 'config.n_lines'
  },
  -- Number of lines within which surrounding is searched
  n_lines = 30,

  -- Duration (in ms) of highlight when calling `MiniSurround.highlight()`
  highlight_duration = 2000,

  -- How to search for surrounding (first inside current line, then inside
  -- neighborhood). One of 'cover', 'cover_or_next', 'cover_or_prev',
  -- 'cover_or_nearest'. For more details, see `:h MiniSurround.config`.
  search_method = 'cover_or_next',
})

-- Remap adding surrounding to Visual mode selection
vim.api.nvim_set_keymap('x', 'S', [[:<C-u>lua MiniSurround.add('visual')<CR>]], { noremap = true })

-- Make special mapping for "add surrounding for line"
vim.api.nvim_set_keymap('n', 'yss', 'ys_', { noremap = false })
```
</details>

As you can see from my config above I'm using the left brackets as a special customization for spaced surroundings (as per your recommendation in the docs) and it works great.

That said, IMO, the below use cases can be improved:

```lua
-- 'ysiw('
test -> ( test )
-- 'cs()' could be improved to remove the spaces, since
-- ) is the no-spacing bind the result should be '(test)'
test -> ( test )
-- 'cs((' results in extra spaces (repeatable)
test -> (  test  )
test -> (   test   )
```

Another thing I noticed is the support for tags with properties:
```lua
-- ysiwtp class="important", results in
<p class="important">test</p class="important">
-- whereas with vim-surround the result would be
<p class="important">test</p>
```
> another side effect of the above invalid tag is that you can no longer use `dst` and `cst` as it fails with an error:
```text
(mini.surround) No surrounding 't' found within 30 lines and `config.search_method = 'cover_or_next'`.
```

**Comment by echasnovski** [comment]
Thanks for the kindest words! I already had a thought that if it was better be named 'many.nvim' :) Didn't anticipate how this project would evolve. Too late now.

Unfortunately, improving the "brackets with space" behavior is impossible under current design. Main obstacle is relying on "balanced item" in Lua patterns (`%b()`, for example) for "input" operations. It can balance only different single character left and right parts. If you are comfortable with finding not balanced pairs, try this setup (with similar values for other brackets):
```lua
['('] = {
  input = { find = '%( .- %)', extract = '^(..).*(..)$' },
  output = { left = '( ', right = ' )' },
},
```

> Another thing I noticed is the support for tags with properties:

That is a good catch! I'll change that.

**Comment by ibhagwan** [comment]
Ty @echasnovski for the quick reply!

>If you are comfortable with finding not balanced pairs

What's the downside of not using balanced pairs? I tried your suggestion on a string such as `( test() )` and it still worked as expected (on the outside pairs).

I'm trying to come up with an `extract` pattern to handle varying amount of spaces, the below almost works, having some issues with a few edge cases, nested brackets `( ( test ) )` cause the error `(mini.surround) Could not extract two surrounding parts. Does your 'config.custom_surroundings.input.extract' pattern has two captures?`

```lua
    ['('] = {
      input = { find = '%( .- %)', extract = '^(.%s+)[^%s]*(%s+.)$' },
      output = { left = '( ', right = ' )' },
    },
```

**Comment by ibhagwan** [comment]
If anyone else ever needs this type of functionality this is my final version, works really well and I can now also use the left brackets to find/replace non spaced surrundings and when I need a balanced pair I use the right bracket.

```lua
    ['('] = {
      input = { find = '%(%s-.-%s-%)', extract = '^(.%s*).-(%s*.)$' },
      output = { left = '( ', right = ' )' },
    },
    ['['] = {
      input = { find = '%[%s-.-%s-%]', extract = '^(.%s*).-(%s*.)$' },
      output = { left = '[ ', right = ' ]' },
    },
    ['{'] = {
      input = { find = '{%s-.-%s-}', extract = '^(.%s*).-(%s*.)$' },
      output = { left = '{ ', right = ' }' },
    },
    ['<'] = {
      input = { find = '<%s-.-%s->', extract = '^(.%s*).-(%s*.)$' },
      output = { left = '< ', right = ' >' },
    },
```

**Comment by echasnovski** [comment]
The tag surrounding should work in latest `main` branch. Thanks again for noticing this!

**Comment by ibhagwan** [comment]
> The tag surrounding should work in latest `main` branch. Thanks again for noticing this!

Thank you @echasnovski, works great!

Btw, I replaced `indent-blankline` with your `indentscope` in [my config](https://github.com/ibhagwan/nvim-lua), your attention to details and perfectionism is admirable - only shows me the relevant scope, column-scope awareness, all perfectly matching the text-objects (I use it for visual mode select) and the animaton just makes it the perfect indent plugin IMO.

**Comment by wodesuck** [comment]
@ibhagwan You can set `find` to `%b()` and keep your `extract` setting, it could match balanced pairs and delete spaces with `ds` and `cs`. The full config would be:
```lua
    ['('] = {
      input = {find = '%b()', extract = '^(.%s*).-(%s*.)$'},
      output = {left = '( ', right = ' )'},
    },
    ['['] = {
      input = {find = '%b[]', extract = '^(.%s*).-(%s*.)$'},
      output = {left = '[ ', right = ' ]'},
    },
    ['{'] = {
      input = {find = '%b{}', extract = '^(.%s*).-(%s*.)$'},
      output = {left = '{ ', right = ' }'},
    },
    ['<'] = {
      input = {find = '%b<>', extract = '^(.%s*).-(%s*.)$'},
      output = {left = '< ', right = ' >'},
    },
```

Btw, since `%s` will match `\n`, the matching surround  may cross lines, and mini.surround doesn't handle this situation yet. I fix it in #114.

**Comment by echasnovski** [comment]
After #103, it will be a more straightforward setup. It will be a default, similar how it is [done in 'mini.ai'](https://github.com/echasnovski/mini.nvim/blob/f8cd711cd59a04ccaecd1178e47ddf292e900cc2/lua/mini/ai.lua#L1051).

**Comment by ibhagwan** [comment]
@wodesuck, fantastic, ty :-)

@echasnovski, ty! watching #103.

**Comment by echasnovski** [comment]
#103 is now resolved after recent updates to `main`. Open brackets now by default include all whitespace into surrounding.

Also, @ibhagwan, format of input surrounding is now soft deprecated in favor of format used in 'mini.ai'. It now has some special meaning to `x.-y` pattern: it now selects *the smallest matched* span possible which is a more symmetric approach regarding search logic. But its downside is that it doesn't really work with with nested matched patterns: in case of `* (  ()  )` with cursor at `*` it will only match inner `()` and not surrounding of `(  ()` (because it is a match with not the smallest width).

All in all, I think new default values are more what people normally need.

**Comment by ibhagwan** [comment]
Ty for the update @echasnovski, just tested, left brackets work great by default.

I do have this customization for lua style strings, is the below still the recommended way of achieving this?
```lua
surround.setup({
  custom_surroundings = {
    S = {
      -- lua bracketed string mapping
      -- 'ysiwS'  foo -> [[foo]]
      input = { find = '%[%[.-%]%]', extract = '^(..).*(..)$' },
      output = { left = '[[', right = ']]' },
    },
  },
}
```

**Comment by echasnovski** [comment]
Yep, [here](https://github.com/echasnovski/mini.nvim/blob/085baefa6aca44526be621b15e7b688f996c2daf/doc/mini.txt#L5399).

Basically, this:

```lua
surround.setup({
  custom_surroundings = {
    S = {
      -- lua bracketed string mapping
      -- 'ysiwS'  foo -> [[foo]]
      input = { '%[%[().-()%]%]' },
      output = { left = '[[', right = ']]' },
    },
  },
}
```

Side note: I would recommend setting this only for 'lua' filetype via `vim.b.minisurround_config`. So with something like this (I put this inside 'after/ftplugin/lua.lua' file):

```lua
vim.b.minisurround_config = {
  custom_surroundings = {
    s = { input = { '%[%[().-()%]%]' }, output = { left = '[[', right = ']]' } },
  },
}
```

**Comment by ibhagwan** [comment]
Ty @echasnovski :heart:
